src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,249 @@
+        {
+            // Throw operation canceled exception before write.
+            ct.ThrowIfCancellationRequested();
+
+            if (isAsync)
+            {
+                await stream.WriteAsync(data.AsMemory(start, length), ct).ConfigureAwait(false);
+            }
+            else
+            {
+                stream.Write(data.AsSpan(start, length));
+            }
+        }
+
+        /// <summary>
+        /// Serializes .NET Decimal to SqlDecimal represented by byte arrays for TDS writing.
+        /// </summary>
+        /// <param name="value">decimal value</param>
+        /// <returns>Byte array for TDS decimal type</returns>
+        public byte[] SerializeDecimal(decimal value)
+        {
+            int[] decimalBits = decimal.GetBits(value);
+            byte[] bytes = _alloc.GetBuffer(17);
+            int current = 0;
+
+            /*
+             Returns a binary representation of a Decimal. The return value is an integer
+             array with four elements. Elements 0, 1, and 2 contain the low, middle, and
+             high 32 bits of the 96-bit integer part of the Decimal. Element 3 contains
+             the scale factor and sign of the Decimal: bits 0-15 (the lower word) are
+             unused; bits 16-23 contain a value between 0 and 28, indicating the power of
+             10 to divide the 96-bit integer part by to produce the Decimal value; bits 24-
+             30 are unused; and finally bit 31 indicates the sign of the Decimal value, 0
+             meaning positive and 1 meaning negative.
+
+             SQLDECIMAL/SQLNUMERIC has a byte stream of:
+             struct {
+                 BYTE sign; // 1 if positive, 0 if negative
+                 BYTE data[]; // 16 bits = int [4]
+             }
+
+             For TDS 7.0 and above, there are always 17 bytes of data
+            */
+
+            // write the sign (note that COM and SQL are opposite)
+            if (0x80000000 == (decimalBits[3] & 0x80000000))
+                bytes[current++] = 0;
+            else
+                bytes[current++] = 1;
+
+            BinaryPrimitives.TryWriteInt32LittleEndian(bytes.AsSpan(current, 4), decimalBits[0]);
+            current += 4;
+            BinaryPrimitives.TryWriteInt32LittleEndian(bytes.AsSpan(current, 4), decimalBits[1]);
+            current += 4;
+            BinaryPrimitives.TryWriteInt32LittleEndian(bytes.AsSpan(current, 4), decimalBits[2]);
+            current += 4;
+            BinaryPrimitives.TryWriteInt32LittleEndian(bytes.AsSpan(current, 4), 0);
+
+            return bytes;
+        }
+
+        /// <summary>
+        /// Dispose Tds Writer
+        /// </summary>
+        public void Dispose()
+        {
+            _alloc.Dispose();
+        }
+
+        /// <summary>
+        /// Disposes Tds Writer and returns a completed <see cref="ValueTask"/>
+        /// </summary>
+        /// <returns>Completed ValueTask</returns>
```

Comment:
Why are we sure that it returns a completed `ValueTask` if the implementation uses `await` and `ConfigureAwait`? Can't it return a task that is still in progress?